packages/grafana-runtime/src/utils/userStorage.ts

```diff
@@ -0,0 +1,130 @@
+import { get } from 'lodash';
+import { lastValueFrom } from 'rxjs';
+
+import { config } from '../config';
+import { BackendSrvRequest, getBackendSrv } from '../services';
+
+const baseURL = `/apis/userstorage.grafana.app/v0alpha1/namespaces/${config.namespace}/user-storage`;
+
+interface RequestOptions extends BackendSrvRequest {
+  manageError?: (err: unknown) => { error: unknown };
+  showErrorAlert?: boolean;
+
+  // rtk codegen sets this
+  body?: BackendSrvRequest['data'];
+}
+
+export type UserStorageSpec = {
+  data: { [key: string]: string };
+};
+
+async function apiRequest<T>(requestOptions: RequestOptions) {
+  try {
+    const { data: responseData, ...meta } = await lastValueFrom(
+      getBackendSrv().fetch<T>({
+        ...requestOptions,
+        url: baseURL + requestOptions.url,
+        data: requestOptions.body,
+      })
+    );
+    return { data: responseData, meta };
+  } catch (error) {
+    return requestOptions.manageError ? requestOptions.manageError(error) : { error };
+  }
+}
+
+function getUserUID(): string {
+  return config.bootData.user.uid === '' ? config.bootData.user.id.toString() : config.bootData.user.uid;
+}
+
+function getResourceName(scope: string): string {
+  return `${scope}:${getUserUID()}`;
+}
+
+function canUseUserStorage(): boolean {
+  return config.featureToggles.userStorageAPI === true && config.bootData.user.isSignedIn;
+}
+
+async function getUserStorage(resourceName: string): Promise<{ spec: UserStorageSpec } | null> {
+  const userStorage = await apiRequest<{ spec: UserStorageSpec }>({
+    url: `/${resourceName}`,
+    method: 'GET',
+    showErrorAlert: false,
+  });
+  if ('error' in userStorage) {
+    if (get(userStorage, 'error.status') !== 404) {
+      console.error('Failed to get user storage', userStorage.error);
+    }
+    // No user storage found, return null
+    return null;
+  }
+  return userStorage.data;
+}
+
+/**
+ * A class for interacting with the backend user storage.
+ */
+export class UserStorage {
+  service: string;
+
+  constructor(service: string) {
+    this.service = service;
+  }
+
+  /**
+   * Retrieves an item from the backend user storage or local storage if not enabled.
+   * @param key - The key of the item to retrieve.
+   * @returns A promise that resolves to the item value or null if not found.
+   */
+  async getItem(key: string): Promise<string | null> {
+    const resourceName = getResourceName(this.service);
+    if (!canUseUserStorage()) {
+      // Fallback to localStorage
+      return localStorage.getItem(resourceName);
+    }
+    const userStorage = await getUserStorage(resourceName);
+    if (!userStorage) {
+      // Also, fallback to localStorage for backward compatibility once userStorageAPI is enabled
+      return localStorage.getItem(resourceName);
+    }
+    return userStorage.spec.data[key];
+  }
```

Comment:
Maybe using IndexedDB might be a closer match to the semantics of a backend API anyway? That would avoid mixing async and sync 'backends' as well?

Reply:
currently, people are already using the `localStorage` for this so I prefer to keep this simple interface so the replacement is as easy as posisble (the only change is the switch from sync to async).

Reply:
I think indexeddb is the better local match for this, but not strongly enough.